server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -254,8 +250,14 @@ public void restoreSnapshot(
                         "snapshot UUID mismatch: expected [" + request.snapshotUuid() + "] but got [" + snapshotId.getUUID() + "]"
                     );
                 }
-                startRestore(repository.getSnapshotInfo(snapshotId), repository, request, repositoryData, updater, l);
-            })), listener::onFailure), listener::onFailure);
+                repository.getSnapshotInfo(
+                    snapshotId,
+                    ActionListener.wrap(
+                        snapshotInfo -> startRestore(snapshotInfo, repository, request, repositoryData, updater, listener),
```

Comment:
Similarly could we add an assertion about the thread on which `startRestore` is running to its precondition checks?

Reply:
jup added to the beginning of `startRestore`